Optimise fault handler assembly

kjbracey opened the pull request:
### Summary of changes <!-- Required -->

Fault handler assembly was very simplistic. Optimise for a 112 byte saving.

Also tidy up formatting, and make all 3 toolchain versions more consistent, for ease of diffing and transferring changes between them.

Also streamline access to crash capture region, avoiding use of pointers. Provides a small saving in fault handler even if crash capture is disabled.

#### Impact of changes <!-- Optional -->



#### Migration actions required <!-- Optional -->


### Documentation <!-- Required -->

None

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [X] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

kjbracey committed:
Optimise fault handler assembly

Fault handler assembly was very simplistic. Optimise for a 112 byte saving.

kjbracey committed:
Reformat except.S

Tidy up formatting, and make all 3 toolchain versions more consistent,
for ease of diffing and transferring changes between them.

0xc0170 commented on platform/source/TARGET_CORTEX_M/TOOLCHAIN_IAR/except.S:
```diff
@@ -42,114 +42,75 @@ FAULT_TYPE_USAGE_FAULT          EQU     0x40
-        LDR     R1,[R3]
-        BL      mbed_fault_handler
+        MOVS    R1,R7
+        SUBS    R1,#21*4
```
what do we do here with substract, can you add a comment? 

Is this similar what we do on line 99 - ADS to R1 value 48

kjbracey replied:
That's rewinding R7 back to the start of the fault context structure. I'm about to change it to just a reload of the address constant anyway - that's simpler once I've killed an unnecessary pointer indirection.

The other one was skipping over saved floating-point registers. (R1 was the stack pointer at that point, here it's preparing the second parameter for mbed_fault_handler)

kjbracey committed:
Streamline crash data access

Use correctly-typed external definition for the crash data region, and
eliminate unnecessary pointer indirection.

Results in a small ROM saving even with crash capture disabled, as there
was a pointer for the fault context store in either case. The pointer
isn't needed, as the context store location is fixed according to the
configuration flag.

0xc0170 commented on platform/source/TARGET_CORTEX_M/TOOLCHAIN_IAR/except.S:
```diff
@@ -42,114 +42,75 @@ FAULT_TYPE_USAGE_FAULT          EQU     0x40
+        MOVS    R1,R6
+        LSRS    R6,R4,#9                ; Check for if STK was aligned by checking bit-9 in xPSR value
+        BCC     Fault_Handler_Continue1
+        ADDS    R1,#0x4
```
Also here: can we comment on this addition or its me who does not see it before going into all instructions prior this one?

kjbracey replied:
That's performing the 8-byte stack alignment referenced (or at least hinted at) in the comment above.

ithinuel commented on platform/internal/mbed_fault_handler.h:
```diff
@@ -48,24 +52,55 @@ typedef struct {
 //This is a handler function called from Fault handler to print the error information out.
 //This runs in fault context and uses special functions(defined in mbed_fault_handler.c) to print the information without using C-lib support.
-void mbed_fault_handler(uint32_t fault_type, const mbed_fault_context_t *mbed_fault_context_in);
+MBED_NORETURN void mbed_fault_handler(uint32_t fault_type, const mbed_fault_context_t *mbed_fault_context_in);
```
Can we make the fault handlers to be marked as no return only in release mode ?
Or at least provide a dead `bx lr` only present when not building in release profile ?
Returning from precise faults has multiple times proven extremely useful.
This is nowadays rather cumbersome to achieve in mbed and I'd like to see it become easier rather than marking more functions as no return.

kjbracey replied:
This would take more than undoing this change - the assembler glue currently doesn't support `mbed_fault_handler` returning, so this was just matching the C annotation to the assembler use. (Get a warning if we do try to return).

Are you thinking more of doing a `BX lr` directly from the fault handler vector to unwind the stack to overcome backtrace unwinding problems in debuggers?

ithinuel replied:
Yes. I often break on `HardFault_Handler` and try to find a `bx lr` somewhere in the flash to then 
```
set $pc = 0x_____
stepi
```
this is most on the time much easier to debug/understand the context of the fault than hopping for the debugger to sort out the content of a potentially corrupted memory.

kjbracey replied:
Another technique is to poke a "BX lr" onto the stack and execute it - there's a script around the place here people use to do that. 

But this is working around a debugger deficiency. There's no fundamental reason in the world why a debugger cannot unwind the stack correctly from the fault vector. (If the BX lr works, the debugger can do the equivalent!)  There are practical issues though, like it having to be in a GDB server, not the GDB core.

I spent a while getting this working a lot better in pyOCD's GDB server, at least, but it would need to be done separately for OpenOCD or any other GDB server. If the pyOCD RTOS awareness is on, then you should get a near-perfect display from the fault handler. https://github.com/mbedmicro/pyOCD/pull/430 would make it basically perfect - showing you the faulting thread up-front without having to select it. I've another PR which would deal with the Thread/Handler stacks separately without RTOS awareness, but that's stale.

Anyway, this has no bearing on `mbed_fault_handler` - by the time you've reached there, you're past the point where a mere `BX lr` would take you back to the fault location. For your technique, stick the breakpoint on `HardFault_Handler`.
